src/box/box.cc

```diff
@@ -2617,8 +2636,8 @@ box_collect_confirmed_vclock(struct vclock *confirmed_vclock, double deadline)
 					      &data.vclock_syncs[replica->id],
 					      deadline) != 0) {
 			/* Timed out. */
-			trigger_clear(&on_ack);
-			trigger_clear(&on_relay_thread_start);
+			while (data.data_access_count > 0)
```

Comment:
It looks like the better solution will be introducing new variable inside relay, which will be reset to false in scope_guard of `box_collect_confirmed_vclock`. We'll check this variable in `relay_trigger_vclock_sync` before accessing `vclock_sync` pointer. @sergepetrenko what do you think?

Reply:
I don't see how we could set this variable.
It's not clear which exact relay is running the `on_relay_thread_start` trigger. And if you try to set the variable in all the relays, it'll lead to a case when other concurrent calls to `box_collect_confirmed_vclock` fail, because the relays beleive they shouldn't set `vclock_sync`.

I think it's better to start allocating the trigger data dynamically and probably reference counting it.

So that you don't have to wait `while (data.data_access_count > 0)` but instead free the data only when `data_access_count` becomes 0. No matter where this happens, right in `box_collect_confirmed_vclock` or `relay_trigger_vclock_sync`.

So basically `relay_trigger_vclock_sync` will always succed setting the `vclock_sync` field, because it will reference the trigger data and it won't be deleted. But as soon as it sets the `vclock_sync`, it will decrement the reference count and might actually delete the data right away.

Reply:
If we have entered this loop, it means that there are currently running `relay_get_sync_on_start`. Here we are waiting for them to finish and wake up our fiber. However, `relay_get_sync_on_start` accepts deadline as one of the arguments. Therefore, everyone we are waiting for here ends when the deadline arrives. So I don't see there any problems.

Reply:
> And if you try to set the variable in all the relays, it'll lead to a case when other concurrent calls to box_collect_confirmed_vclock fail, because the relays beleive they shouldn't set vclock_sync.

Yeah, sounds reasonable, not great.

> But as soon as it sets the vclock_sync, it will decrement the reference count and might actually delete the data right away.

I like the idea of allocating `sync_trigger_data` dynamically. AFAICS, freeing cannot be done in `box_collect_confirmed_vclock`, as it may exit earlier, then trigger is executed. Should be implemented inside `relay_trigger_vclock_sync`, when counter is 0, we free the memory.

Reply:
I suggest switching to refcounts and dynamic allocation.

Waiting until some other code stops using our stack-allocated data doesn't look too good.